labs/projecteuler/projecteuler.md

```diff
@@ -0,0 +1,132 @@
+```
+Clearly we don't need SAT solvers to figure out how to subtract one.  And writing simple factorization code in Python factored 3,000,013 in 0.419s.  But there is something really interesting here: to solve the problem, instead of figuring out how to write code which would solve the problem for us, we just had to write the problem in Cryptol code and let a SAT solver solve it for us!  Usually it's much easier to write a problem than write the solution.  We can't always expect the SAT solver to find the answer quickly (this is known as as the [P vs. NP problem](https://en.wikipedia.org/wiki/P_versus_NP_problem)), but there are many times where it will come back with answer quickly.  We like to say that it doesn't cost much to ask a SAT solver.
+
+## Project Euler Problems
+
+[Project Euler](https://projecteuler.net/) is a collection of mathematical problems that can be solved using relatively simple computer code (i.e. to do them, you shouldn't need to create optimized code to run on a supercomputer).  Some of the questions lend themselves to querying a SAT solver.  We've collected them here for you to try your hand at.
+
+ONE IMPORTANT NOTE: we have reworded the questions to make them more Cryptol-friendly.  The Project Euler questions often ask you to find all the numbers which satisfy some property and then combine them in some way (e.g. take their collective sum).  For most of the problems, it is hard enough to get Cryptol to find an answer, let alone all possible answers.  For these exercises, it is enough to write code that will find an answer, not all of them.
+
+### [Problem 9](https://projecteuler.net/problem=9)
+
+> A Pythagorean triplet is a set of three natural numbers, a < b < c, for which a^2 + b^2 = c^2.
+> 
+> For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.  There exists exactly one Pythagorean triplet for which a + b + c = 1000.  Find this triple.
+
+### [Problem 34](https://projecteuler.net/problem=34)
+
+> 145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.
+>
+> Find all numbers which are equal to the sum of the factorial of their digits.
+> 
+> Note: as 1! = 1 and 2! = 2 are not sums they are not included.
+>
+> (Aside: these numbers are called [factorions](https://en.wikipedia.org/wiki/Factorion).  It is unclear why Project Euler does not include 1 and 2).
+>
+> Hints: the factorial function is usually defined recursively, but that tends to make SAT solving difficult.  Since you only need to calculate the factorial of the numbers 0-9, make your function just do a case by case calculation.  To get the digital representation of the number, create a function which takes in a number and a list of numbers and returns true exactly when the list is the base 10 representation.  Finally, it can be shown that the most number of di\
+gits a factorion can have is 6.
+
+### [Problem 36](https://projecteuler.net/problem=36)
+
+> The decimal number, 585 = 1001001001_(2) (binary), is palindromic in both bases.
```

Comment:
The representation of the binary is strange.  Perhaps rephrase it as "The decimal number 585 and its binary representation 1001001001 are both [palindromic](https://www.dictionary.com/browse/palindromic)."